src/Strings/ValidParanthesis.java

```diff
@@ -0,0 +1,210 @@
+					else {
+						aster_count--;
+					}
+				}
+			}
+			return true;
+		}
+		else {
+			return false;
+		}
+	}
+	public static boolean checkValidStringAL(String s) {
+		char[] char_arr = s.toCharArray();
+		if(char_arr.length == 0) {
+			return true;
+		}
+		ArrayList<Character> track_par = new ArrayList<Character>();
+		for (int i=char_arr.length-1; i >= 0 ; i--) {
+			if (char_arr[i] == ')' || char_arr[i] == '*') {
+				track_par.add(0, char_arr[i]);
+			}
+			// The case for left paranthesis
+			else {
+				// If the ( paranthesis occurs and then there is no correspoding )/*. Then we return false.
+				if (track_par.isEmpty()) {
+					return false;
+				}
+				else {
+					int initial_size = track_par.size();
+					for (int j=0; j<initial_size; j++) {
+						if (track_par.get(j) == ')') {
+							track_par.remove(j);
+							break;
+						}
+					}
+					// This happens when no ) was found
+					if (initial_size == track_par.size()) {
+						track_par.remove(0);
+					}
+				}
+			}
+		}
+		// If the stack is empty everything matched, or if the stack has asterisk
+		if (track_par.isEmpty()) {
+			return true;
+		}
+		// If there is a left over close paranthesis, then we have to fail
+		else if (track_par.get(0) == ')') {
+			return false;
+		}
+		else if (track_par.get(0) == '*') {
+			int aster_count = 0;
+			for (int i = 0; i<track_par.size(); i++) {
+				if (track_par.get(i) == '*') {
+					aster_count++;
+				}
+				else {
+					// If it has no prior *
+					if (aster_count == 0) {
+						return false;
+					}
+					// If there is a right paranthesis, decrement the * count
+					else {
+						aster_count--;
+					}
+				}
+			}
+			return true;
+		}
+		else {
+			return false;
+		}
+	}
+	/**
+	 * VM
+	 * Using 2 stacks
+	 * Time Complexity - O(N + S) -> S being stack size, if it is negligible compared to N -> O(N)
+	 * Space Complexity - O(N)
+	 * **/
+	public static boolean checkValidStringWithTwoStacks(String s) {
+		Stack<Integer> openStack = new Stack();
+		Stack<Integer> starStack = new Stack();
+
+		for(int i = 0; i < s.length(); i++){
+			char c = s.charAt(i);
+			// place all indices of ( in open stack
+			if(c == '(')
+				openStack.push(i);
+			// place all indices of * in star stack
+			else if(c == '*')
+				starStack.push(i);
+			else{
+				// when we see a ) -  look at the open stack to check for corresponding ( we saw earlier
+				if(!openStack.isEmpty()){
+					openStack.pop();
+				} else {
+					// if open stack was empty - look at the star stack for substituting it as (
+					if(!starStack.isEmpty())
+						starStack.pop();
+					else
+						// if both are empty - there is no ( or * in front of )
+						return false;
+				}
+			}
+		}
+		// at the end - if we are still left with (
+		// use the star stack to substitute * as )
+		// if you can't find a * as substitute - false
+		while(!openStack.isEmpty()){
+			if(!starStack.isEmpty()){
+				if(starStack.pop() < openStack.pop())
```

Comment:
This check is to see that the * appears only after a ( is used as a substitute ). i.e the index of * should not be lesser than index of (

Reply:
You are doing an ascii comparison here? Why is it needed. Wouldn’t this always be false. * has a higher ascii value than open parenthesis always.

Reply:
I think your logic here is pop from both the stacks. You don’t need an if statement for this. It should work without the if.

Reply:
The stacks store the index where I saw a ( bracket and *. It is the index comparison, I am not storing the char as such, so it is not ascii comparison. 

The if statement on the pop is to see that the index of ( is lesser than index of *, which means there is always a star somewhere after the ( to nullify it. If the star is before (, it won't be able to nullify it.

Reply:
Yeah got it. Sorry got confused thinking you were pushing the character. Index makes sense.

Reply:
he is doing integer comparision right? the index.